keystorage/yubihsm2/src/main/java/tech/pegasys/signers/yubihsm/YubiHsmSession.java

```diff
@@ -0,0 +1,503 @@
+      throws YubiHsmException {
+    if (response.length < MESSAGE_MAC_SIZE) {
+      throw new IllegalArgumentException(
+          "Response is too short to contain a " + MESSAGE_MAC_SIZE + " bytes MAC");
+    }
+    final byte[] respMac =
+        Arrays.copyOfRange(response, response.length - MESSAGE_MAC_SIZE, response.length);
+    final byte[] respMacMsg = Arrays.copyOfRange(response, 0, response.length - MESSAGE_MAC_SIZE);
+    final byte[] fullResponseMac = getMac(sessionRMacKey, challenge, respMacMsg);
+    final byte[] rmac = Arrays.copyOfRange(fullResponseMac, 0, MESSAGE_MAC_SIZE);
+    if (!Arrays.equals(rmac, respMac)) {
+      throw new YubiHsmException("Incorrect MAC");
+    }
+  }
+
+  /**
+   * Extracts the encrypted command response from the device's response to SessionMessage
+   *
+   * @param rawResponse The device response to SessionMessage
+   * @return The device's encrypted response to the command inside SessionMessage
+   * @throws YubiHsmException If the session ID returned by the device does not match this session
+   *     ID or if the device had returned an error code
+   */
+  private Bytes getSessionMessageResponse(final Bytes rawResponse) {
+    Bytes resp = CommandUtils.parseCmdResponse(Command.SESSION_MESSAGE, rawResponse);
+    if (resp.get(0) != sessionID) {
+      throw new YubiHsmException("Incorrect session ID");
+    }
+
+    // extract the response to the command
+    // Response: 1 byte session ID + response to the command + 8 bytes MAC
+    return resp.slice(1, resp.size() - MESSAGE_MAC_SIZE - 1).copy();
+  }
+
+  @Override
+  public void close() {
+    LOG.trace("Status: {}", status);
+
+    if (status != SessionStatus.CREATED && status != SessionStatus.AUTHENTICATED) {
+      LOG.debug("Session is not open. Doing nothing");
+      return;
+    }
+
+    try {
+      LOG.trace("Closing session {}", sessionID);
+      final Bytes response = sendSecureCmd(Command.CLOSE_SESSION, Bytes.wrap(new byte[0]));
+      CommandUtils.verifyResponseLength(Command.CLOSE_SESSION, response.size(), 0);
+    } catch (final YubiHsmException e) {
+      if (Objects.equals(e.getYubiHsmError(), Errors.INVALID_SESSION)) {
+        LOG.info("YubiHSM Session {} is not valid", sessionID);
+      } else {
+        LOG.warn("Unexpected error while closing session", e);
+      }
+    } finally {
+      destroySessionKeys();
+      status = SessionStatus.CLOSED;
+    }
+  }
+
+  private void destroySessionKeys() {
+    if (sessionEncKey != null) {
+      Arrays.fill(sessionEncKey, (byte) 0x00);
+    }
+    if (sessionMacKey != null) {
+      Arrays.fill(sessionMacKey, (byte) 0x00);
+    }
+    if (sessionRMacKey != null) {
+      Arrays.fill(sessionRMacKey, (byte) 0x00);
+    }
+    LOG.trace("Destroyed the session encryption key, MAC key and RMAC key");
+  }
+
+  // visible for testing
+  Bytes getHostChallenge() {
+    return Bytes.wrap(new SecureRandom().generateSeed(HALF_CHALLENGE_SIZE));
```

Comment:
do we want to generate a new SecureRandom() every time this is called, or should we do it once somewhere, and pass it in? Just to protect our entropy.